### DISCUSSION TITLE
Strange effect when using pre-releases

### DISCUSSION THREAD

**Discussion by goto40** [callowayproject/bump-my-version#243] [Q&A] [answered]
When I adopt the manual to use pre-releases with v0.27.0 (thanks for that feature! :-) ):
```
[tool.bumpversion.parts.pre_l]
values = ["dev", "rc", "final"]
optional_value = "final"
```

but I just want "dev" to be present (no "rc", "final" and no optional_value)
```
[tool.bumpversion.parts.pre_l]
values = ["dev"]
```

I get an unexpected behavior (at least unexpected to me).

Instead of 
```
$ bump-my-version show-bump
0.1.0 ── bump ─┬─ major ─ 1.0.0-dev0
               ├─ minor ─ 0.2.0-dev0
               ├─ patch ─ 0.1.1-dev0
               ├─ pre_l ─ invalid: The part has already the maximum value among ['dev', 'rc', 'final'] and cannot be bumped.
...
```

I then get
```
$ bump-my-version show-bump
0.1.0 ── bump ─┬─ major ─ 1.0.0
               ├─ minor ─ 0.2.0
               ├─ patch ─ 0.1.1
...
```

Is this intentional?

**Comment by coordt** [comment, marked as answer]
This is the expected behavior.

Here is some background

1. **There is a difference between the values of all the parts and how they are serialized into a version.** In the example in the docs, version `0.1.0` could also be serialized as `0.1.0-final0`. The `pre_l` value `final` is optional as is the `pre_n` value of `0`, and therefore means they doesn't have to be in the serialized version.
2. **There is always an optional value.** The first item in a list of values is the optional value, unless you indicate otherwise. In your alteration for `pre_l` and setting it to `["dev"]`, you have made `dev` optional. You have also made it the only allowed value.
3. **Bump-my-version tries to selected the serialization method that requires the fewest components.** The docs contain an [explanation](https://callowayproject.github.io/bump-my-version/explanation/mental-model/#how-a-version-is-serialized) of how it selects a a serialization value.

If what you want is a releasing strategy that looks something like:

- 1.0.0-dev0
- 1.0.0-dev1
- 1.0.1

You would change the configuration of the `pre_l` component to only contain `["dev", "final"]`

Was I able to answer your question?

**Reply by goto40** [reply]
Thank you very much for your quick reply! I need to take time to discuss your answer. I will reply next 10 days.

**Reply by goto40** [reply]
I was missing the fact that there is always one optional part. Thank you!
